# Homework Introduction to Python

$\textbf{Ubah nama di bawah yaa!}$

- Nama : Radisha Fanni Sianti
- Batch : 31

In [18]:
import numpy as np

# please import as much as you need

# Normal

## 1. Operasi Variabel

> Misal kita memiliki toko sepatu. Modal membeli 1 kodi sepatu adalah Rp 1.000.000. Setiap sepatu dijual seharga Rp 80.000. Buatlah variabel modal, kuantitas, dan harga_jual sesuai informasi di atas. Hitung profit yang didapatkan jika semua sepatu berhasil terjual, dan masukkan ke variabel profit.

Saran langkah-langkah pengerjaan
1. Convert terlebih dahulu kodi dalam jumlah barang (1 kodi = xyz buah sepatu)
2. Gunakan operasi matematika sederhana untuk menghitung profit menggunakan tiga buah variabel yang sudah ditentukan

In [37]:
# your code goes here!
modal = 1000000
kuantitas = 20
harga_jual = 80000
profit = (kuantitas * harga_jual ) - modal
print("Profit yang didapatkan jika semua sepatu berhasil terjual adalah", profit)

Profit yang didapatkan jika semua sepatu berhasil terjual adalah 600000


Jadi, dari profit di atas, dapat disimpulkan bahwa profit yang didapatkan jika semua sepatu berhasil terjual adalah 600000

## 2. Package Python dan Dictionary

> Kita akan coba mengakses data scientific article dari PLOS corpus
    Menggunakan API request dengan kode berikut : 
```
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
```
**perhatikan tanda kutipnya tidak salah ketika meng-copy ke dalam python

Pertanyaannya : 
Convert response.content dari request tersebut dari JSON Format ke dalam Dictionary, dan print-kan key di dalamnya apa saja

Saran langkah-langkah pengerjaan
1. Ikuti code dari soal, kemudian ekstrak value dari variable response menggunakan dot (.) content menjadi sebuah variabel baru (dictionary)
2. Gunakan library json, convert hasil content menggunakan json.loads
3. Print keys dari dictionary tersebut


In [9]:
# your code goes here!
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
print(response)

import json
dictionary = response.content
dictionary_real = json.loads(dictionary)
print(dictionary_real)
print("Keys pada dictionary pertama adalah ",dictionary_real.keys())
print("Keys pada dictionary response adalah ",dictionary_real["response"].keys())
print("Keys pada dictionary docs adalah ",dictionary_real["response"]["docs"][0].keys())

<Response [200]>
{'response': {'numFound': 1501, 'start': 0, 'docs': [{'id': '10.1371/journal.pone.0026564', 'abstract': ['\n        Little is known about the viruses infecting most species. Even in groups as well-studied as Drosophila, only a handful of viruses have been well-characterized. A viral metagenomic approach was used to explore viral diversity in 83 wild-caught Drosophila innubila, a mushroom feeding member of the quinaria group. A single fly that was injected with, and died from, Drosophila C Virus (DCV) was added to the sample as a control. Two-thirds of reads in the infected sample had DCV as the best BLAST hit, suggesting that the protocol developed is highly sensitive. In addition to the DCV hits, several sequences had Oryctes rhinoceros Nudivirus, a double-stranded DNA virus, as a best BLAST hit. The virus associated with these sequences was termed Drosophila innubila Nudivirus (DiNV). PCR screens of natural populations showed that DiNV was both common and widespread 

Jadi, dari url link di atas, kami menyimpulkan bahwa ata scientific article dari PLOS corpus memiliki keys response, numFound, start, docs, id dan abstract

# Intermediate

## 3. Operasi dalam List

> Buatlah list dengan 10 nama kota di indonesia, kemudian:
1. Urutkan nama kota di dalam list berdasarkan abjad (a)
2. Ambil 4 kota pertama dari list tersebut (yang telah diurutkan) (b)
3. Ambil 3 kota terakhir dari list tersebut (yang telah diurutkan) (c)

Saran langkah-langkah pengerjaan
1. Tidak ada ketentuan nama kota (misal berdasarkan provinsi X atau pulau Y), dibebaskan untuk memilih. Buat nama-nama kota tersebut dalam sebuah list.
2. Gunakan fungsi sorted(X) untuk mengurutkan secara kecil-ke-besar (sesuai abjad), dimana X adalah contoh variabel list pada langkah (a)
3. Gunakan hasil poin (a), urutkan nama kota di dalam list berdasarkan abjad, untuk mengerjakan poin (b) dan (c). 

In [15]:
# your code goes here!
nama_kota = ["Aceh", "Medan", "Palembang", "Lampung","Jakarta", 
             "Bandung","Semarang","Yogyakarta","Surabaya","Malang"]
#1
print("Urutan mama kota berdasarkan abjad adalah", sorted(nama_kota),"\n")

#2
print("4 Kota pertama dari list adalah ",sorted(nama_kota)[0:4],"\n")

#3
print("3 Kota terakhir dari list adalah ",sorted(nama_kota)[-3:])

Urutan mama kota berdasarkan abjad adalah ['Aceh', 'Bandung', 'Jakarta', 'Lampung', 'Malang', 'Medan', 'Palembang', 'Semarang', 'Surabaya', 'Yogyakarta'] 

4 Kota pertama dari list adalah  ['Aceh', 'Bandung', 'Jakarta', 'Lampung'] 

3 Kota terakhir dari list adalah  ['Semarang', 'Surabaya', 'Yogyakarta']


Jadi, dari operasi list di atas, dapat disimpulkan bahwa :

Urutan mama kota berdasarkan abjad adalah ['Aceh', 'Bandung', 'Jakarta', 'Lampung', 'Malang', 'Medan', 'Palembang', 'Semarang', 'Surabaya', 'Yogyakarta'] 

4 Kota pertama dari list adalah  ['Aceh', 'Bandung', 'Jakarta', 'Lampung'] 

3 Kota terakhir dari list adalah  ['Semarang', 'Surabaya', 'Yogyakarta']

## 4. Function dalam List

> Team sales memiliki list total transaksi dari 10 merchant pada bulan Juni 2020, yaitu: 
```
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]
```
1. Trx minimum
2. Rata-rata 
3. Trx maksimal

Saran langkah-langkah pengerjaan
1. Copy variabel total_transaksi yang ada di soal ke notebook kita (atau ketik ulang boleh)
2. Kemudian, hitung menggunakan function operasi dalam list yang telah dipelajari di kelas pada topik list. Print hasilnya setiap poin (a), (b), dan (c)

In [36]:
# your code goes here!
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]

#1
print ("Total transaksi minimum adalah ",np.min(total_transaksi))

#2
print ("Rata-rata total transaksi adalah ", np.sum(total_transaksi)/len(total_transaksi))

#3
print ("Total transaksi maksimal adalah ",np.max(total_transaksi))


Total transaksi minimum adalah  10
Rata-rata total transaksi adalah  25.6
Total transaksi maksimal adalah  45


Jadi, dari hasil operasi list di atas, dapat disimpulkan bahwa nilai minimal, max, rata-rata adalah sebagai berikut :

Total transaksi minimum adalah  10

Rata-rata total transaksi adalah  25.6

Total transaksi maksimal adalah  45

# Advanced

## 5. NumPy Array

> Buatlah array dari 5 bilangan ganjil pertama (dimulai dari 1), Kemudian lakukan perintah berikut : 
1. Kali-kan array tersebut dengan 10
2. Bagi-kan array (hasil dari bagian a) dengan 3
3. Kali-kan array (hasil dari bagian b) dengan array baru :
`[-45, 43, 3, 56, 5]`
4. Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c) 

Saran langkah-langkah pengerjaan
1. HATI-HATI dalam pengerjaan, karena apabila poin sebelumnya salah, maka selanjutnya akan kebawa salah (misal (b) salah, maka poin (c) sampai (d) akan salah)
2. Untuk array 5 bilangan ganjil pertama boleh manual boleh menggunakan fungsi, dimulai dari 1
3. Ikuti perintah soal, gunakan numpy array, jangan list. Sesuaikan dengan permintaan soal pada setiap butir.



In [24]:
# your code goes here!
bilangan_ganjil = np.array([1,3,5,7,9])

#1
hasil1 = bilangan_ganjil*10
print ("a. Hasil kali array dengan 10 adalah ", hasil1, "\n")

#2
hasil2 = hasil1/3
print ("b. Hasil bagi array a dengan 3 adalah ", np.round(hasil2,2),"\n")

#3
hasil3 = hasil2 * np.array([-45,43,3,56,5])
print ("c. Hasil kali array b dengan array baru adalah", np.round(hasil3,2),"\n")

#4
print("Indeks tertinggi array c adalah ", np.argmax(hasil3),"\n")
print("Indeks terendah array c adalah ", np.argmin(hasil3))

a. Hasil kali array dengan 10 adalah  [10 30 50 70 90] 

b. Hasil bagi array a dengan 3 adalah  [ 3.33 10.   16.67 23.33 30.  ] 

c. Hasil kali array b dengan array baru adalah [-150.    430.     50.   1306.67  150.  ] 

Indeks tertinggi array c adalah  3 

Indeks terendah array c adalah  0


Jadi, dari hasil operasi numpy di atas, dapat disimpulkan bahwa :

a. Hasil kali array dengan 10 adalah  [10 30 50 70 90] 

b. Hasil bagi array a dengan 3 adalah  [ 3.33 10.   16.67 23.33 30.  ] 

c. Hasil kali array b dengan array baru adalah [-150.    430.     50.   1306.67  150.  ] 

Indeks tertinggi array c adalah  3 

Indeks terendah array c adalah  0

# Terimakasih, salam dari Surabaya!